[ENH] Added Builder and execute options for `page` and added execution timeout key

AakashGfude opened the pull request:
`page` command will now have
-  All the execution options which can be specified in the `_config.yml` file. 
-  An addition `--builder` option to specify the builders 
-  -W option for warning as error

Have also added the capability to specify timeout `_config.yml` inside the `execute` object with `timeout` key.

Should have made two separate PRs :sweat_smile: . Anyways, the change needed for timeout is quite less.

WIll make a commit to update the docs in this PR.

cc : @choldgraf @chrisjsewell 

fixes #741

choldgraf commented on jupyter_book/commands/__init__.py:
```diff
@@ -117,106 +117,61 @@ def build(path_book, path_output, config, toc, warningiserror, builder):
+    "--builder",
+    default="html",
+    help="Which builder to use. Must be one of {BUILDER_OPTIONS}",
 )
```
where did the execute option go? Am I missing it?

AakashGfude replied:
Because we have all the options for execute now, like `cache`, `force`, `off`  in the `_config` file, so it did not make sense to have on and off option here.

choldgraf commented on jupyter_book/commands/__init__.py:
```diff
@@ -399,3 +337,103 @@ def init(path, kernel):
+
+
+# utility functions
+def builder_specific_actions(exc, builder, output_path, cmd_type, page_name=None):
```
please document this function so it's easier for others to maintain

AakashGfude replied:
@choldgraf thanks, I will document it

choldgraf commented:
hey @AakashGfude - thanks for the improvements. Some short comments in there and I can try to give this a deeper look later.

One thought I have: it seems like the diff for this PR is much bigger than I'd expect given that it's just adding a few options to the `page` function. Can you describe the other changes you've made?

In addition - I don't know that we want to force people to use `_config.yml` to build with pages...is there a way that we can recommend people control the behavior purely from the command line. For pages in particular, I'd expect people may not want to deal with `_config.yml`.

Finally - it seems you've done a bunch of rearranging in order to have shared functionality between `build` and `page`. This makes me wonder - should we deprecate the `page` CLI option entirely, and instead add it as a special case of the `build` verb? E.g., if `build` points to a *folder*, then assume it is a jupyter book. If build points to a *file*, then assume it is a single page and follow that logic instead. Curious if that could reduce some of the complexity in the codebase

AakashGfude commented:
> One thought I have: it seems like the diff for this PR is much bigger than I'd expect given that it's just adding a few options to the `page` function. Can you describe the other changes you've made?
> 
@choldgraf  apart from creating a common function for builder options and a few adjustments in the page command, I have created a new `test_page` file, removed the testing of page from `test_build` and added some tests for pdf in `test_pdf` .



> In addition - I don't know that we want to force people to use `_config.yml` to build with pages...is there a way that we can recommend people control the behavior purely from the command line. For pages in particular, I'd expect people may not want to deal with `_config.yml`.
> 
Hmm. There will be a few options we will need to incorporate for page then in the command line, like `execute method`, `timeout`, `builder` type. It won't be too long I think, because the user might not use each option all the time and only the defaults. So, I think you are right, we should give user freedom to not deal with `_config.yml` . 
How about we have both? If they want to configure more, they can use the `--config` option, else if just want to change one of these options use command-line argument.  Because if we want it to be as powerful as build, we might want the `--config` for future as well.

> Finally - it seems you've done a bunch of rearranging in order to have shared functionality between `build` and `page`. This makes me wonder - should we deprecate the `page` CLI option entirely, and instead add it as a special case of the `build` verb? E.g., if `build` points to a _folder_, then assume it is a jupyter book. If build points to a _file_, then assume it is a single page and follow that logic instead. Curious if that could reduce some of the complexity in the codebase

This idea definitely stands out. We anyways have planned to make their functionality similar. In the case of single pages, we just need to not allow `toc` , and have to set some config beforehand like `to_exclude`; else most of them are essentially the same. This should reduce some complexity for the user having to use a separate option for essentially the same thing and also in the codebase.  I think this is a good option.

choldgraf commented:
@AakashGfude would you like to explore a single `build` command before merging this? Or do you wanna get this one merged and *then* explore a single build command?

AakashGfude commented:
@choldgraf I will explore a single build command before merging this.

AakashGfude commented:
Hey @choldgraf , have merged build and page. Let me know what you think. Will edit the documentation once this looks okay to you.

choldgraf commented on jupyter_book/commands/__init__.py:
```diff
@@ -41,7 +41,7 @@ def main():
 
 @main.command()
-@click.argument("path-book")
+@click.argument("path-src")
```
can we make the user-facing thing a full word? (`path-source`) we can rename to a shorter name in the code (`path_src`) if we want

AakashGfude replied:
Ooh. my bad. Will do that. Thanks

AakashGfude commented:
Hey @choldgraf , in my local system, the tests are passing. But in the CI, it is throwing `unsupported theme option 'number_toc_sections' given`. Is it because of an old theme copy either in my machine or the CI? It happened after taking a pull from master though.

choldgraf commented:
@AakashGfude ah yeah that's because of some updates I made to the sphinx-book-theme and I haven't yet updated jupyter-book to take them in :-/  is it OK if we wait to get those changes in so that we can rebase and make sure the tests are all OK? I'll try to do it today or tomorrow

choldgraf commented:
just merged the patch in - wanna rebase and see if that fixes the tests?

AakashGfude commented:
@choldgraf , I think the `book_template` folder in build does not have `intro.md` file? Hence, pdfhtml is failing?
![image](https://user-images.githubusercontent.com/6542997/87781757-a64c4800-c874-11ea-9d9b-9ddcab505d99.png)

choldgraf commented:
restarting this to see if the new release of SBT will fix the failing circle job. let's see!

AakashGfude commented:
Hey @choldgraf , trying to debug the issue, I think the theme still might have some issues. As the previous version of my PR, where it was working is also giving the error:- 

![image](https://user-images.githubusercontent.com/6542997/88364687-0a1bc700-cdc7-11ea-808c-b4a76b91b503.png)

Have pinned it down to a function in `jinja2glue.py` file of sphinx. Will continue digging.

AakashGfude commented:
@choldgraf so, I have narrowed the issue in `sphinx-book-theme` repo to this commit :- https://github.com/executablebooks/sphinx-book-theme/commit/f5ea24db4831f044153b6dfe40222c0310061be7 . Before this commit the build is working fine.

choldgraf commented:
hmmm - I *bet* the issue is here:

https://github.com/executablebooks/sphinx-book-theme/commit/f5ea24db4831f044153b6dfe40222c0310061be7#diff-95c3c14f442154a7a77462f1e36e81e7R62

that's where we're calling the `sphinx` toctree function, and maybe it is `None` for some pages?

AakashGfude commented:
Cool @choldgraf. I will have a play around with it later and see if I can pinpoint.

AakashGfude commented:
@choldgraf the build throws an error before the function `generate_nav_html` is called, where https://github.com/executablebooks/sphinx-book-theme/commit/f5ea24db4831f044153b6dfe40222c0310061be7#diff-95c3c14f442154a7a77462f1e36e81e7R62 is present. 

The error occurs after this event call is finished :- 

`app.connect("html-page-context", add_to_context)`

More specifically in the following line :-  `return concat(self.root_render_func(self.new_context(vars)))` inside the render function of `jinja2` package. 

The `concat` function here specifically throws the error. 

Since, there are quite a few changes in the commit https://github.com/executablebooks/sphinx-book-theme/commit/f5ea24db4831f044153b6dfe40222c0310061be7 , it is turning out to be a bit time consuming to pinpoint the issue.
Let me know if anything strikes you.

choldgraf commented:
I figured out what was wrong in the sphinx-book-theme, fixed it, and just pushed a new release (thanks for finding a bug in SBT by the way ;-) ). Tests should now be passing!

Are we ready to 🚢?

AakashGfude commented:
Thanks @choldgraf . Somehow missed that message. Have updated the files, according to the latest commits

choldgraf commented on docs/start/overview.md:
```diff
@@ -28,13 +28,13 @@ If you'd like to help with adding Windows support, please [say hello in this iss
 Building a Jupyter Book broadly consists of two steps:
 
-1. **Put your book content in a folder**. Jupyter Book needs the following
+1. **Put your book content in a folder or a file**. Jupyter Book needs the following
```
```suggestion
1. **Put your book content in a folder**. Jupyter Book needs the following
```

I think that for the "getting started" section we should still just focus on a book use-case, and underneath we can add a `{note}` block that mentions you can build individual files as well. what do you think?

AakashGfude commented:
> For example, when building [quantecon-example](https://github.com/executablebooks/quantecon-example) when building the file `book/docs/matplotlib.md` it places a `_build` folder in `book/docs` however when building the project as a whole the `_build` folder is contained inside of the `book` folder.
> This will cause issues as `.gitignore` typically is setup for a single `_build` location and it would be more tidy if there was a global `_build` folder. What do you think?

@mmcky true it would be tidier. but the present approach ensures separation of builds. A whole folder build will be in a different directory and individual page builds will be in a different directory. This can prevent overlaps like when someone creates a page build. And then goes on to do a project build, the page build will be merged with the folder if in same directory. 

Do you think it is an important feature to have? separate distinguishable build files for folder and page? We can also create subfolders in `_builld` for folder and page.

mmcky commented:
> This can prevent overlaps like when someone creates a page build.

I see. So a `page` build is different to a `project` build in general with regards to the `html`. I agree -- in that case we do want to keep them separate. I think though perhaps `_build/page/<builder>` might be a better location in that case for builds that parse `page` only.

chrisjsewell commented:
Perhap we could have a `.jbook` folder in the top-level of the repo, and store all build folders there?

mmcky commented:
> Perhap we could have a `.jbook` folder in the top-level of the repo, and store all build folders there?

I like this from a `dev` perspective but myconcern here is that on a lot of computers a `.` folder is not browsable through `finder` etc. So many users may find it hard to open `build` files.

AakashGfude commented:
> I see. So a `page` build is different to a `project` build in general with regards to the `html`? In that case I agree -- we do want to keep them separate. I think though perhaps `_build/_page/<builder>` might be a better location for builds that parse `page` only items.
> 
> I am just a bit worried about a proliferation of `_build` folders and also ensures `jb clean` picks up all `_builds`

That's an important point. To be able to clean all folder with `jb clean`. I agree should be a subset of the same outdir. So, you are suggesting for single page builds have, for example `_build/_page/html/<file>` for HTML builds

AakashGfude commented:
> Perhap we could have a `.jbook` folder in the top-level of the repo, and store all build folders there?

perhaps having a name with `build` or `output` or something will make it more obvious that it is the output of the build?

chrisjsewell commented:
maybe `_jbuild`?

mmcky commented:
> So, you are suggesting for single page builds have, for example _build/_page/html/<file> for HTML builds

Yeah -- that is what I was thinking. If you build lots of `individual` pages they would all go into that folder as `unique` and independent builds next to each other. If you built a `pdflatex` of a single page then it would go to `_build/_page/latex`. 

> perhaps having a name with build or output 

I think the `_` helps identify it is a `built` folder by some software. But that may be my dev hat on.

AakashGfude commented:
> Yeah -- that is what I was thinking. If you build lots of `individual` pages they would all go into that folder as `unique` and independent builds next to each other. If you built a `pdflatex` of a single page then it would go to `_build/_page/latex`.

sounds good 

> I think the `_` helps identify it is a `built` folder by some software. But that may be my dev hat on.

yes, I meant with an `_`. was just mentioning having the substring `build`, `output` etc somewhere in the name. :)

mmcky commented:
> maybe `_jbuild`?

I don't mind a generic `_build` myself unless you think it is confusing with `sphinx` projects?

chrisjsewell commented:
It just makes it more obbious that it has been created by jupyter-book; most users of jupyter-book wouldn;t really know (or care) about sphinx conventions, thats just an implementation detail

mmcky commented:
> _jbuild just makes it more obvious that it has been created by jupyter-book;

roger that -- I don't really mind either way on `_build` or `_jbuild`. I agree I think most users won't know about sphinx so wasn't too worried by generic `_build`. 

@AakashGfude perhaps lets update this PR without changing the base `_build` folder (but moving `page` builds to be included in the global `_build` folder) and then we can open a separate PR modifying the base `_build` to use `_jbuild`. That way we will get clear consensus on the change in the default `outdir` name through a single PR.

mmcky commented:
@AakashGfude ready for me to test this again. Please ping me once I should go ahead. Thanks.

AakashGfude commented:
@mmcky yes ready. have a go at this.

chrisjsewell commented:
Heya, is this ready to merge now/soon?
Happy to merge with failing tests, before I update myst-parser/myst-nb versions to fix that, since this PR overlaps with what I want to add during that upgrade; in terms of configuration options, that include the timeout key.

I'll document that option, along with the others. So don't worry about doing that here 😄

mmcky commented:
@chrisjsewell Want to do some testing of this PR today to make sure assumptions around where the build folder are robust. @AakashGfude indicated the outdir metadata was not available at the moment of setting up for page build so had to make some assumptions around that folder placement.

mmcky commented:
@chrisjsewell are you saying you need the `execution timeout` code removed from this PR or you'd just like it merged?

chrisjsewell commented:
> @chrisjsewell are you saying you need the `execution timeout` code removed from this PR or you'd just like it merged?

Either way is fine 😄 I just want to "use" it in my PR, so either its already there, or I have to write it in

mmcky commented:
@AakashGfude I am trying to checkout your branch locally but I haven't seen this error before. 

```
▶ gh pr checkout 744
remote: Enumerating objects: 137, done.
remote: Counting objects: 100% (137/137), done.
remote: Compressing objects: 100% (26/26), done.
remote: Total 122 (delta 79), reused 122 (delta 79), pack-reused 0
Receiving objects: 100% (122/122), 16.14 KiB | 1.47 MiB/s, done.
Resolving deltas: 100% (79/79), completed with 12 local objects.
From https://github.com/executablebooks/jupyter-book
 ! [rejected]          refs/pull/744/head -> page-and-timeout  (non-fast-forward)
exit status 1
```
any ideas?

mmcky commented on jupyter_book/commands/__init__.py:
```diff
@@ -57,218 +59,101 @@ def main():
+    config_overrides = {}
+    freshenv = False
+    BUILD_PATH = (
+        path_output if path_output is not None else find_config_path(PATH_SRC_FOLDER)
```
@AakashGfude so the assumption here is that if path_output is `none` then it calls a `utility` function to search for the `config` file and assume the `_build` folder is in that directory. 

Does `path_output` come from the `default` values in `config` from `jupyter_book`? It looks like the old code had

```
BUILD_PATH = path_output if path_output is not None else PATH_BOOK
```

and I assume `path_book` would be specified from the `CLI` by a user.

AakashGfude replied:
@mmcky path_output is something which the user can explicitly specify in the command line while building , and `PATH_BOOK` is the source folder for the build

AakashGfude replied:
https://github.com/executablebooks/jupyter-book/blob/master/jupyter_book/commands/__init__.py#L63  . like here in the master branch, it just took the path of `path_book` which is the source folder

mmcky replied:
Right -- so won't they be the `default` values for the `_build` folder should it exist.  I guess I don't understand why we don't keep using those default paths for setting the `BUILD_PATH`?

AakashGfude replied:
righ @mmcky , that should be the default value for `_build` in case of `book` but, in case of nested `page`  we want it to be in the parent folders, where the `_build` of `book` would have resided.

AakashGfude replied:
@mmcky do you want me to move this line inside the `if` statement of `page` only, and put the previously used line there again?. I understand the confusion.

mmcky replied:
Isn't `_page` just a subfolder of the main `_build` folder such as `_build/_page/` or are you building a `_build` and a `_page` at the same level?

AakashGfude replied:
but the basic functionality of that function is that it returns the `source_path` unless there is any `_config.yml` in the parent folders, in which case it will return the path to that folder. 

So, in the case of `book` it seems full proof to me. 
But in the case of `page` if the parent `book` folder does not have a `_config.yml` file, then the function won't be able to determine the correct folder to create the `_build` for `page` (i,e, the folder where the `_build` for `book` is), and it will default to the path of the `page` only, as the path of `_build`.

For this case, I think the code needs a lot of rearranging and time, so we can create an issue and I can start rearranging code later. But for now, we can merge this PR?

mmcky replied:
ok thanks for explaining. Happy to raise an issue and loop back to this later if needed. I haven't been able to test locally -- but @AakashGfude has done some testing. So happy for this to be merged.

AakashGfude commented:
> any ideas?

did it work for you @mmcky ?

chrisjsewell commented:
> Want to do some testing of this PR today to make sure assumptions around where the build folder are robust. 

Just doing a few other bits then I can have a look if you want.
Otherwise, I trust your judgement

mmcky commented:
> > any ideas?
> 
> did it work for you @mmcky ?

No still can't pull the PR locally using `gh` cmd

AakashGfude commented:
> No still can't pull the PR locally using `gh` cmd. Maybe it is because of the `force-push`
@mmcky yes that might be true, I will push an empty commit now and see if it works

mmcky commented:
> No still can't pull the PR locally using `gh` cmd. Maybe it is because of the `force-push`
> @mmcky yes that might be true, I will push an empty commit now and see if it works

Negative :-(. it is saying `non-fast-forward`
Maybe: https://stackoverflow.com/questions/4684352/what-does-git-push-non-fast-forward-updates-were-rejected-mean

mmcky commented:
ah maybe you need to fetch upstream?

AakashGfude commented:
@mmcky its up-to-date with upstream master. but I had used all the destructive commands mentioned in that link like, `git commit --amend`, `git rebase` and `force-push` :sweat_smile: .

mmcky commented:
ok -- I guess I need to fetch your repo then.

mmcky commented:
ok - thanks @AakashGfude for this PR. I have done some testing and just have a few comments.

1. The folder structure seems to work well in `_build/_page`. I see you setup a `folder` with the same `document` name to hold the build files. I was thinking `_build/_page/start/overview.html` would be the path but I understand why you need a folder here. My only thought is if you have two `overview.md` source files in different subfolders you would get a collision. In my mind there are two use cases for building files directly 1. build simple single source files easily, and 2. preview a page in a project. The second use case may cause this collision without the `source` directory structure. 

2. One behaviour I think we need to document (perhaps a simple `reminder` or `note`) is if users try to build documents that have links to other documents in their project they will get build warnings and incomplete links. For example, building `docs/start/overview.md` will issue a bunch of `unknown document`, `term not in glossary`, and `undefined links` warnings. This is expected behaviour -- as the purpose of page is mainly to produce single page documents.

AakashGfude commented:
> 1. The folder structure seems to work well in `_build/_page`. I see you setup a `folder` with the same `document` name to hold the build files. I was thinking `_build/_page/start/overview.html` would be the path but I understand why you need a folder here. My only thought is if you have two `overview.md` source files in different subfolders you would get a collision. In my mind there are two use cases for building files directly 1. build simple single source files easily, and 2. preview a page in a project. The second use case may cause this collision without the `source` directory structure.

yes, thought of creating this `folder` to have defined separate `HTML`, `latex` files for each page. 
Collision is a good point, and I think this folder structure would prevent that if we append the folder name with its subfolder as well. What do you think? 

> 2. One behaviour I think we need to document (perhaps a simple `reminder` or `note`) is if users try to build documents that have links to other documents in their project they will get build warnings and incomplete links. For example, building `docs/start/overview.md` will issue a bunch of `unknown document`, `term not in glossary`, and `undefined links` warnings. This is expected behaviour -- as the purpose of page is mainly to produce single page documents.

Should we write in the docs where we documented the single page build, as a warning?

mmcky commented:
> thought of creating this folder to have defined separate HTML, latex files for each page.

Good thinking.

> I think this folder structure would prevent that if we append the folder name with its subfolder

This convention would also work. You mean `_build/_page/{dir}-{filname}/html/` It would also avoid too much nested folders. You may end up with long folder names if the file is heavily nested but I don't think that is a likely scenario in the real world. 

> Should we write in the docs where we documented the single page build, as a warning?

Just a simple `reminder` would be good I think -- for those using this pathway to build "preview" documents in a project

AakashGfude commented:
@mmcky have added `_build/_page/{dir}-{filname}/html/` and a `reminder` doc. Feel free to edit the `note` in the doc.

mmcky commented on docs/start/build.md:
````diff
@@ -46,25 +46,24 @@ Sometimes you'd like to build a single page of content rather than an
 Your page will be called `mypage.html`. This will work
 for any {doc}`content source file <../content-types/index>` that is supported by Jupyter Book.
 
+```{note}
````
@AakashGfude I don't have push access but we could simplify it to this. 

```md
Users should note that building **single pages** in the context of a larger project can trigger warnings and incomplete links. For example, building `docs/start/overview.md` will issue a bunch of `unknown document`,
 `term not in glossary`, and `undefined links` warnings. 
```

then we can 🚀

mmcky commented:
@AakashGfude just retesting the latest commits. 

I am getting

```
Desktop/jupyter-book/docs  page-and-timeout ✔                                                                                        36m  
▶ jb build advanced/pdf.md 
Usage: jb build [OPTIONS] PATH_BOOK
Try 'jb build --help' for help.

Error: Invalid value for 'PATH_BOOK': Directory 'advanced/pdf.md' is a file.
```

mmcky commented:
Thanks @AakashGfude -- @chrisjsewell I think this is good to go if you're happy with `CI` results.

mmcky commented:
@AakashGfude can you take a look at the failing test. I got the same `pytest` result on my local `scratch` environment -- I think it is related to the `test_pdfhtml` option. It looks like CI should be up and running again with the latest releases

mmcky reviewed (approved):
Thanks for iterating on this @AakashGfude

AakashGfude commented:
Thanks for testing this out @mmcky

chrisjsewell commented:
Eurgh @AakashGfude what is this awful merge commit you've done 😒 
Where's the adherence to the https://github.com/executablebooks/.github/blob/master/CONTRIBUTING.md#commit-messages
Just let me know next time you want to merge a PR

mmcky commented:
hey @AakashGfude -- I have been letting @chrisjsewell be the gatekeeper for merging PR's on the `jupyter-book` projects. So let's let him merge once he is happy with them. 

hey @chrisjsewell would you like us to revert this and submit as a clean PR?

chrisjsewell commented:
Yeh its not the end of the world, but it does hurt my heart to see a messy commit history lol

chrisjsewell commented:
I believe revering just adds to the commit history.
I can rebase master to clean it up, but this would affect any forked branches containing these commits, so I will need to go through the PRs that have been updated since the merge and also rebase them: I think just #888 #895 #723

chrisjsewell commented:
Actually I don't think any PRs have incorporated these commits yet, so I am going to go ahead and do it 🤞

mmcky commented:
Let us know if we can help with the tidy up.

chrisjsewell commented:
Yeh @mmcky could you write here a summary for the commit.
I was going to use the initial comment of this PR, but I think it has changed since then

chrisjsewell commented:
I'm going to squash it all into one commit, since it I don't see any natural "groupings" in the current commits

chrisjsewell commented:
> Is there a way to make Squash and Merge the default on github?

I've turned off the ability to merge commit 😄

chrisjsewell commented:
Going from the diff, it looks like there is no longer any changes to the actual CLI, i.e. no additional flags.
But now you can specify a single file, in addition to a directory, for builds.
Plus the addition of the `execute.timeout` `_congig.yml` option

mmcky commented:
@chrisjsewell how is:

---

This PR merges the `page` and `build` commands to enable `jb build` to build both `projects` and `single documents`. 

It also adds:
-  `execute.timeout` option which can be specified in the `_config.yml` file. 

fixes #741

chrisjsewell commented:
ah yes thanks, I see it remove the `page` command as well 👍

chrisjsewell commented:
Ok done! b40ea84495e8cf282895273109e4373af4e25674

For future reference, I did this by:

- `git rebase -i e916c57fad6c03f1c09e83a12a43d31bc35148b7`
- `git push --force`

As mentioned it very "dangerous" to do this on public branches, that people are forking from.
But for your own PR forks, if you have multiple "chuncks" of commits that you would like to keep separate in the final merge, you can just do `git rebase -i master` + `git push --force`

The interactive rebase provides you the list of commits, since the specified initial commit, and you can change how they are processed; mainly using the `fixup` and `reword` tags

mmcky commented:
Nice history log and `git` sleuthing.

chrisjsewell commented:
Ok bear with me two secs, while we are at it, I have added a commit before this one, to pin (correctly) pin myst-nb to v0.8.
and now I am going to release a v0.7.5 from that commit (before this PR), so that people are getting myst-nb v0.9 from jupyter-book v0.7, which are not strictly compatible 😬

mmcky commented:
The joy of a `stack` -- glad to have you managing this upgrade @chrisjsewell

chrisjsewell commented:
Yep and the importance of proper version pinning!

chrisjsewell commented:
Right! Done another little reshuffle (thats the last!) and pinned attrs, so that the tests now pass 🎉 
Then I have released v0.7.5 to PyPi for the commit before this PR (see https://github.com/executablebooks/jupyter-book/commits/master)

So now this PR and updating to myst-nb v0.9 can go in a "breaking" v0.8.0 release

mmcky commented:
Nicely done @chrisjsewell

chrisjsewell commented:
I just hope it doesn't break anyone's forks/PRs 😬

AakashGfude commented:
@mmcky @chrisjsewell sorry guys :'(, not a good day to take a break for an appointment. Thanks, @chrisjsewell for handling my inexperience well. A good lesson for me to not do this, and also how to handle it if this mistake happens.

phaustin commented:
> But for your own PR forks, if you have multiple "chuncks" of commits that you would like to keep separate in the final merge, you can just do `git rebase -i master` + `git push --force`
> 
> The interactive rebase provides you the list of commits, since the specified initial commit, and you can change how they are processed; mainly using the `fixup` and `reword` tags

re: interactive rebase -- I've found this vscode extension to be very helpful:  https://marketplace.visualstudio.com/items?itemName=trentrand.git-rebase-shortcuts

credit: https://stuartleeks.com/posts/working-with-git-rebase-in-visual-studio-code/
